packages/lodestar/src/db/api/beacon/repository.ts

```diff
@@ -0,0 +1,84 @@
+import {IBeaconConfig} from "../../../config";
+import {IDatabaseController} from "../../controller";
+import {Bucket, encodeKey} from "../../schema";
+import BN from "bn.js";
+import {AnySSZType, deserialize, hashTreeRoot, serialize} from "@chainsafe/ssz";
+
+export type Id = Buffer | string | number | BN;
+
+export abstract class DatabaseRepository<T> {
+  protected config: IBeaconConfig;
+
+  protected db: IDatabaseController;
+
+  protected bucket: Bucket;
+
+  protected type: AnySSZType;
+
+  protected constructor(
+    config: IBeaconConfig,
+    db: IDatabaseController,
+    bucket: Bucket,
+    type: AnySSZType) {
+    this.config = config;
+    this.db = db;
+    this.bucket = bucket;
+    this.type = type;
+  }
+
+  public async get(id: Id): Promise<T | null> {
+    try {
+      return deserialize(await this.db.get(encodeKey(this.bucket, id)), this.type);
+    } catch (e) {
+      return null;
+    }
+  }
+
+  public async has(id: Id): Promise<boolean> {
+    return !! await this.get(id);
+  }
+
+  public async storeUnderRoot(value: T): Promise<void> {
+    await this.store(hashTreeRoot(value, this.type), value);
+  }
+
+  public async store(id: Id, value: T): Promise<void> {
+    await this.db.put(encodeKey(this.bucket, id), serialize(value, this.type));
+  }
+
+  public async delete(id: Id): Promise<void> {
+    await this.db.delete(encodeKey(this.bucket, id));
+  }
+
+}
+
+export abstract class FullDatabaseRepository<T> extends DatabaseRepository<T> {
```

Comment:
Maybe `Repository` and `BulkRepository` for the names. DatabaseRepository feels redundant somehow.